hamocc/mo_dmsph.F90

```diff
@@ -0,0 +1,145 @@
+
+    use mo_control_bgc, only: io_stdo_bgc
+    use netcdf,         only: nf90_noerr,nf90_nowrite,nf90_close,nf90_open
+    use mod_xc,         only: mnproc,xchalt
+
+    implicit none
+    INTEGER, INTENT(in) :: kpie,kpje
+    INTEGER ::i,j,l
+
+    REAL,intent(in) ::omask(kpie,kpje)
+
+    ! define the fields
+    REAL :: pi_ph_in(kpie,kpje,pi_ph_record)
+    INTEGER ncid,ncstat
+
+    ! allocate pi_ph field
+    if(.not. allocated(pi_ph)) call alloc_pi_ph(kpie,kpje)
+
+    !
+    ! Open netCDF data file
+    !
+    IF(mnproc==1) THEN
+       ncstat = NF90_OPEN(trim(pi_ph_path)//'MONTHLY_PI_PH.nc',        &
+     &                   NF90_NOWRITE, ncid)
+       write(io_stdo_bgc,*) 'HAMOCC: opening MONTHLY_PI_PH file'
+       IF (ncstat.NE.NF90_NOERR ) THEN
+          CALL xchalt('(get_pi_ph: Problem with netCDF1)')
+                 stop '(get_pi_ph: Problem with netCDF1)'
+       END IF
+    END IF
+    !
+    ! Read  data
+    call read_netcdf_var(ncid,'pH',pi_ph_in(1,1,1),pi_ph_record,0,0)
+    !
+    ! Close file
+    IF(mnproc==1) THEN
+       ncstat = NF90_CLOSE(ncid)
+       IF ( ncstat .NE. NF90_NOERR ) THEN
+          CALL xchalt('(get_pi_ph: Problem with netCDF200)')
+                 stop '(get_pi_ph: Problem with netCDF200)'
+       END IF
+    END IF
+
+    ! set missings over land
+    do l=1,pi_ph_record
+       do j=1,kpje
+          do i=1,kpie
+             if(omask(i,j).gt.0.5) then
+                pi_ph(i,j,l) = pi_ph_in(i,j,l)
+             else
+                pi_ph(i,j,l) = 0.
+             endif
+          enddo
+       enddo
+    enddo
+
+  END SUBROUTINE GET_PI_PH
+
+
+  !**********************************************************************
+  ! Allocate the PI_PH field.
+  !**********************************************************************
+  subroutine alloc_pi_ph(kpie,kpje)
+    use mod_xc,         only: mnproc
+    use mo_control_bgc, only: io_stdo_bgc
+
+    implicit none
+    integer, intent(in) :: kpie,kpje
+    integer             :: errstat
+
+    IF (mnproc.eq.1) THEN
+       WRITE(io_stdo_bgc,*)'Memory allocation for variable pi_ph ...'
+       WRITE(io_stdo_bgc,*)'First dimension    : ',kpie
+       WRITE(io_stdo_bgc,*)'Second dimension   : ',kpje
+       WRITE(io_stdo_bgc,*)'Third dimension    : ',pi_ph_record
+    ENDIF
+
+    ALLOCATE (pi_ph(kpie,kpje,pi_ph_record),stat=errstat)
+    if(errstat.ne.0) stop 'not enough memory pi_ph'
+    pi_ph(:,:,:) = 0.0
+
+  end subroutine alloc_pi_ph
+
+
+  !**********************************************************************
+  ! Get DMS_PH as a function of pH
+  !**********************************************************************
+  function get_dmsph(i,j,kplmon) result(dms_ph)
+    use mo_carbch, only: hi
+
+    implicit none
+
+    integer, intent(in) :: i,j,kplmon
+    real :: dms_ph
+
+    dms_ph  = 1. + (-log10(hi(i,j,1)) - pi_ph(i,j,kplmon))*dms_gamma
+
+  end function get_dmsph
```

Comment:
I this one line should be moved back to ocprod once pi_ph is passed to ocprod

Reply:
I'm not sure why we would need to pass `pi_ph` to `ocprod`. Currently I set `pi_ph` as `private`, and I think all the references to this field can be internal to `mo_dmsph`.

Reply:
A problem with passing pi_ph through the argument list is that it may not be allocated, and probably we don't want to allocate memory for pi_ph unless we actually use it.

Reply:
Yes, that is true, that would be the price to pay. Regarding memory I wouldn't see this as a big issue since it is a 2d field. So  if `with_phdms=.false.`, we would pass a 2d-field set to 0.0 by `get_ph_clim` (or whatever the name is)